docs/design.md

```diff
@@ -0,0 +1,423 @@
+index.
+
+### Persistence and reconciliation
+
+Indices will be persisted in a filesystem directory, referenced by their shard
+key. This makes for easy identification.
+
+Although not expected, it is possible that over time the index repo and the shard management layer state will diverge (entropy):
+
+1. Active shards may have their indices deleted (e.g. if an operator deleted a
+   file or the directory).
+2. Destroyed shards may have their indices lying around (e.g. if there was some
+   system error in housekeeping).
+
+A method `ReconcileIndices()` can be used to reconcile the index repo with the
+shard management layer, by:
+
+1. Fetching gone indices from their mounts.
+2. Deleting lingering indices.
+
+### Interface
+
+```go
+type IndexRepo interface {
+   FullIndexRepo
+   ManifestRepo
+}
+
+type FullIndexRepo interface {
+   // public
+   GetFullIndex(key shard.Key) (FullIndex, error)
+
+   // private, called only by shard management when registering
+   // and destroying shards
+   InsertFullIndex(key shard.Key, index FullIndex) error
+   DeleteFullIndex(key shard.Key) (bool, error)
+}
+
+type ManifestRepo interface {
+   // public
+   GetManifest(key ManifestKey) (Manifest, error)
+   InsertManifest(key ManifestKey, manifest Manifest) error
+   DeleteManifest(key ManifestKey) (bool, error)
+   ExistsManifest(key ManifestKey) (bool, error)
+}
+
+type ManifestKey struct {
+   Shard    key.Shard
+   Rule     string
+   Version  string
+}
+
+type FullIndex interface {
+   Offset(c cid.Cid) (offset int64, err error) // inexistent: offset=-1, err=nil
+   Contains(c cid.Cid) (bool, error)
+   Len() (l int64, err error)
+   ForEach(func(c cid.Cid, offset int64) (ok bool, err error)) error
+}
+```
+
+## DAG access layer
+
+This layer is responsible for serving DAGs or sub-DAGs from one or many shards.
+Initially, queries will require the shard key. That is, queries will be bounded
+to a single identifiable shard.
+
+In the future, when the cross-shard top-level index is implemented, the DAG
+store will be capable of resolving the shard for any given root CID.
+
+DAG access layer allows various access abstractions:
+   1. Obtaining a Blockstore bound to a single shard.
+   2. Obtaining a Blockstore bound to a specific set of shards.
+   3. Obtaining a global Blockstore.
+   4. `ipld.Node` -- TBD.
+
+At this point, we are only concerned with (1). The remaining access patterns
+will be elaborated on in the future.
+
+```go
+type DAGAccessor interface {
+   Shard(key shard.Key) ShardAccessor
+}
+
+type ShardAccessor interface {
+   // Close() the accessor, signalling to the DAG store that we're done here and
+   // possible transient copies can be reclaimed once the shard is released.
+   io.Closer
+
+   Blockstore() (ReadBlockstore, error)
+}
+```
+
+## Requirements for CARv2
+
+- Index needs to be detachable.
+- Index offsets need to be relative to the CARv1, and not absolute in the
+  physical CARv2 file.
+- Index needs to be iterable.
```

Comment:
Index is a map; is there a specific API you have in mind for this?

Reply:
We were hoping that the index would not have to be loaded into memory (which a map is). Something like this would work: https://github.com/filecoin-project/dagstore/pull/2/files#diff-b0929dacd4f1fc192cd308c1a9db2e85154ea742ccec315518e659d7f6a3efc7R347

Reply:
@masih

The Carbs Index API/impl is a good reference:

We also need the ability to Marshal/Unmarshal Indexes so we can persist them to disk/read them from disk if need be.

Reply:
Captured: https://github.com/ipld/go-car/issues/95